src/platforms/hosted/bmp_libusb.c

```diff
@@ -19,13 +19,72 @@
 
 /* Find all known usb connected debuggers */
 #include "general.h"
+#if defined(_WIN32) || defined(__CYGWIN__)
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#include <stdint.h>
+
+#include "ftd2xx.h"
+#else
 #include <libusb.h>
+#include "libftdi1/ftdi.h"
+#endif
 #include "cli.h"
 #include "ftdi_bmp.h"
 #include "version.h"
+#include "probe_info.h"
 
 #define NO_SERIAL_NUMBER "<no serial number>"
 
+void get_bmp_product_version_string(struct libusb_device_descriptor *device_descriptor, libusb_device *device,
+	libusb_device_handle *handle, char **product, char **manufacturer, char **serial, char **version);
+
+typedef struct debugger_device {
+	uint16_t vendor;
+	uint16_t product;
+	bmp_type_t type;
+	void (*function)(
+		struct libusb_device_descriptor *, libusb_device *, libusb_device_handle *, char **, char **, char **, char **);
+	char *type_string;
+} debugger_device_s;
+
+//
+// Create the list of debuggers BMDA works with
+//
+debugger_device_s debugger_devices[] = {
+	{VENDOR_ID_BMP, PRODUCT_ID_BMP, BMP_TYPE_BMP, get_bmp_product_version_string, "Black Magic Probe"},
+	{VENDOR_ID_STLINK, PRODUCT_ID_STLINKV2, BMP_TYPE_STLINK_V2, NULL, "ST-Link v2"},
+	{VENDOR_ID_STLINK, PRODUCT_ID_STLINKV21, BMP_TYPE_STLINK_V2, NULL, "ST-Link v2.1"},
+	{VENDOR_ID_STLINK, PRODUCT_ID_STLINKV21_MSD, BMP_TYPE_STLINK_V2, NULL, "ST-Link v2.1 MSD"},
+	{VENDOR_ID_STLINK, PRODUCT_ID_STLINKV3_NO_MSD, BMP_TYPE_STLINK_V2, NULL, "ST-Link v2.1 No MSD"},
+	{VENDOR_ID_STLINK, PRODUCT_ID_STLINKV3, BMP_TYPE_STLINK_V2, NULL, "ST-Link v3"},
+	{VENDOR_ID_STLINK, PRODUCT_ID_STLINKV3E, BMP_TYPE_STLINK_V2, NULL, "ST-Link v3E"},
+	{VENDOR_ID_SEGGER, PRODUCT_ID_UNKNOWN, BMP_TYPE_JLINK, NULL, "Segger JLink"},
+	{VENDOR_ID_FTDI, PRODUCT_ID_FTDI_FT2232, BMP_TYPE_FTDI, NULL, "FTDI FT2232"},
+	{VENDOR_ID_FTDI, PRODUCT_ID_FTDI_FT4232, BMP_TYPE_FTDI, NULL, "FTDI FT4232"},
+	{VENDOR_ID_FTDI, PRODUCT_ID_FTDI_FT232, BMP_TYPE_FTDI, NULL, "FTDI FT232"},
+	{0, 0, BMP_TYPE_NONE, NULL, ""},
+};
+
+bmp_type_t get_type_from_vid_pid(uint16_t probe_vid, uint16_t probe_pid)
+{
+	bmp_type_t probe_type = BMP_TYPE_NONE;
+	for (size_t index = 0; debugger_devices[index].type != BMP_TYPE_NONE; index++) {
+		/*
+		 * Segger probe PIDs are unknown, if we have a Segger probe force the type to JLink
+		 */
+		if (probe_vid == VENDOR_ID_SEGGER) {
+			probe_type = BMP_TYPE_JLINK;
+			break;
+		}
+		if (probe_vid == debugger_devices[index].vendor && probe_pid == debugger_devices[index].product) {
+			probe_type = debugger_devices[index].type;
+			break;
```

Comment:
```suggestion
			return debugger_devices[index].type;
```
(and remove the then needless braces)

Reply:
Edits as suggested